hwdb.d/60-seat.hwdb

```diff
@@ -0,0 +1,36 @@
+# This file is part of systemd.
+#
+# This file lists graphic devices that don't have a DRM driver and fall back to
+# a frame-buffer one instead. Since commit 6260d28b8a, frame-buffer devices are
+# no more considered as graphical capable.
+#
+# The matches have the 'fb' prefix to make sure that only the framebuffer
+# device, and not the (parent) PCI device, is tagged with 'master-of-seat'.
+#
+# Allowed properties are:
+#    ID_TAG_MASTER_OF_SEAT=1
+
+#########################################
+# eXtreme Graphic Innovation (XGI)
+#########################################
+
+# Z7/Z9 (XG20 core), a rather ancient graphic chip, doesn't have a DRM driver
+# and falls back to vesafb.
+fb:pci:v000018CAd00000020*
+ ID_TAG_MASTER_OF_SEAT=1
+
+#########################################
+# HyperV
+#########################################
+
+# HyperV currently doesn't do DRM, hence we need to synthesize for HyperV's fb
```

Comment:
There is support from v5.14: https://github.com/torvalds/linux/commit/76c56a5affeba1e163b
Does this interfere with using that? Or is it unrelated?

Reply:
I guess this is related, however the compat entry in hwdb is still needed for a while until kernels < 5.14 are no more supported.

Reply:
these are the perils of encoding facts about software rather than about hardware in hwdb...

Reply:
An udev rule was more flexible in this regard… But let's wait for 5.14 to be released. If it turns out that this is any problem, we can add some rule that looks at the kernel version or whatnot. We don't need to solve this now.

Reply:
Can we postpone this to v250 then? Autumn distros might use v5.14 and systemd v249

Reply:
Maybe we should pass the version of the kernel that get DRM support for the specific HW from hwdb and use that in the udev rule to decide whether the framebuffer needs `master-of-seat` tag or not.

Reply:
> Can we postpone this to v250 then? Autumn distros might use v5.14 and systemd v249

I don't think so. First, at this point we don't know if 5.14 will be released with the driver, or maybe it'll get reverted or moved to staging or whatever. If if passes that first hurdle, we don't know if various distros will enable it by default. OTOH, this patch fixes real issues, so we want it sooner rather than later.

> Maybe we should pass the version of the kernel that get DRM support for the specific HW from hwdb and use that in the udev rule to decide whether the framebuffer needs master-of-seat tag or not.

Maybe. How would that look exactly?

Is it an issue if we have this rule and a working driver? Maybe it's just a bit ugly but harmless otherwise?

Reply:
> Maybe. How would that look exactly?

Dunno maybe by introducing a new feature in udev (builtin or whatever), not sure whether it's worth the trouble though.

> Is it an issue if we have this rule and a working driver? Maybe it's just a bit ugly but harmless otherwise?

We would reintroduce issue #10435 for these devices. But given the fact that GDM got some improvements that make it less subject to this issue, having this rule and a DRM driver wouldn't be an issue I would say.

Reply:
We wouldn't because incidentally in 5.14 it is planned for vesafb to also get a DRM capable driver. When this happens all of this drm/non-drm split pretty much goes away, at least for common hardware.

Reply:
Will we need to change this at that point? Or is it transparent?

Reply:
At that point the device will get flagged twice - once as the framebuffer corresponding to a specific known device, once as DRM capable. So in theory it should just work.

Reply:
ah that's great to hear, thanks!

Reply:
nah, if both drivers exist in parallel things will be racy, we need to drop the entry from the hwdb then. The problem is that it's not clear if the drm or fb device will show up first, and thus graphics apps that want to use drm preferably and fall back to fb otherwise might sometimes end up with the wrong device. If that wasn#t an issue then we could tag all fb devices out of the box, and in fact we used to do that. We dropped that only in order to fix the race. 

I mean, if there are userpace drivers for these things I guess we could just tell these packages to ship hwdb entires/udev rules for the devices they care about and then drop them from upstream. i.e. it's a property of certain software and hence should be shipped with said software. The big issue with that is just that vesafb is kinda a catchall, i.e. devices that have oo drivers of their own, hence no obvious candidate package to ship those rules.

Reply:
The problem is that vesafb which is not DRM capable shows up first, and binding to it blocks the real driver from working.

With DRM support in vesafb this problem goes away when the display server uses DRM with vesafb it can switch to the real driver once available.

Recognizing that vesafb is DRM capable and using the DRM interface for it is another separate problem, and probably one that needs to be handled by the display server. Same problem exists for all DRM capable drivers.